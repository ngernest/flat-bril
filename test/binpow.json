{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "acc",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "x",
            "n",
            "acc"
          ],
          "dest": "y",
          "funcs": [
            "bin_pow"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "x",
            "two"
          ],
          "dest": "half",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "half",
            "two"
          ],
          "dest": "reconstruct",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x",
            "reconstruct"
          ],
          "dest": "same",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "same"
          ],
          "op": "ret"
        }
      ],
      "name": "is_even",
      "type": "bool"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "acc",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "n",
            "zero"
          ],
          "dest": "cond0",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "cond0"
          ],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "args": [
            "acc"
          ],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "n"
          ],
          "dest": "is_even",
          "funcs": [
            "is_even"
          ],
          "op": "call",
          "type": "bool"
        },
        {
          "args": [
            "is_even"
          ],
          "labels": [
            "then.1",
            "else.1"
          ],
          "op": "br"
        },
        {
          "label": "then.1"
        },
        {
          "args": [
            "x",
            "x"
          ],
          "dest": "x2",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "n2",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "x2",
            "n2",
            "acc"
          ],
          "dest": "val",
          "funcs": [
            "bin_pow"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "val"
          ],
          "op": "ret"
        },
        {
          "label": "else.1"
        },
        {
          "args": [
            "n",
            "one"
          ],
          "dest": "n2",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "x",
            "acc"
          ],
          "dest": "acc2",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x",
            "n2",
            "acc2"
          ],
          "dest": "val",
          "funcs": [
            "bin_pow"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "val"
          ],
          "op": "ret"
        }
      ],
      "name": "bin_pow",
      "type": "int"
    }
  ]
}
