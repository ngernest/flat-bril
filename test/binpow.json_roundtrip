{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "acc",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "x",
            "n",
            "acc"
          ],
          "dest": "y",
          "funcs": [
            "bin_pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "y"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "x",
            "two"
          ],
          "dest": "half",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "half",
            "two"
          ],
          "dest": "reconstruct",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x",
            "reconstruct"
          ],
          "dest": "same",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "same"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "is_even",
      "type": "bool"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "acc",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "two",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "n",
            "zero"
          ],
          "dest": "cond0",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "cond0"
          ],
          "funcs": [],
          "labels": [
            "then.0",
            "else.0"
          ],
          "op": "br"
        },
        {
          "label": "then.0"
        },
        {
          "args": [
            "acc"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.0"
        },
        {
          "args": [
            "n"
          ],
          "dest": "is_even",
          "funcs": [
            "is_even"
          ],
          "labels": [],
          "op": "call",
          "type": "bool"
        },
        {
          "args": [
            "is_even"
          ],
          "funcs": [],
          "labels": [
            "then.1",
            "else.1"
          ],
          "op": "br"
        },
        {
          "label": "then.1"
        },
        {
          "args": [
            "x",
            "x"
          ],
          "dest": "x2",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "n",
            "two"
          ],
          "dest": "n2",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "x2",
            "n2",
            "acc"
          ],
          "dest": "val",
          "funcs": [
            "bin_pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "val"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.1"
        },
        {
          "args": [
            "n",
            "one"
          ],
          "dest": "n2",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "x",
            "acc"
          ],
          "dest": "acc2",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "x",
            "n2",
            "acc2"
          ],
          "dest": "val",
          "funcs": [
            "bin_pow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "val"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "bin_pow",
      "type": "int"
    }
  ]
}
