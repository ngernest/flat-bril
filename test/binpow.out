@main
	op: const		dest: acc		type: int		value: 1	
	op: call 		dest: y 		type: int		args: ["x", "n", "acc"]		funcs: ["bin_pow"]
	op: print		args: ["y"]	
@is_even
	op: const		dest: two		type: int		value: 2	
	op: div  		dest: half		type: int		args: ["x", "two"]	
	op: mul  		dest: reconstruct		type: int		args: ["half", "two"]	
	op: eq   		dest: same		type: bool		args: ["x", "reconstruct"]	
	op: ret  		args: ["same"]	
@bin_pow
	op: const		dest: zero		type: int		value: 0	
	op: const		dest: one		type: int		value: 1	
	op: const		dest: two		type: int		value: 2	
	op: eq   		dest: cond0		type: bool		args: ["n", "zero"]	
	op: br   		args: ["cond0"]		labels: ["then.0", "else.0"]
.then.0
	op: ret  		args: ["acc"]	
.else.0
	op: call 		dest: is_even		type: bool		args: ["n"]		funcs: ["is_even"]
	op: br   		args: ["is_even"]		labels: ["then.1", "else.1"]
.then.1
	op: mul  		dest: x2		type: int		args: ["x", "x"]	
	op: div  		dest: n2		type: int		args: ["n", "two"]	
	op: call 		dest: val		type: int		args: ["x2", "n2", "acc"]		funcs: ["bin_pow"]
	op: ret  		args: ["val"]	
.else.1
	op: sub  		dest: n2		type: int		args: ["n", "one"]	
	op: mul  		dest: acc2		type: int		args: ["x", "acc"]	
	op: call 		dest: val		type: int		args: ["x", "n2", "acc2"]		funcs: ["bin_pow"]
	op: ret  		args: ["val"]	
